torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3385,11 +3385,40 @@ def _make_data_dependent_error(self, expr, unhinted_expr):
             # problem
         )
 
-    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr") -> None:
+    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr", msg: str) -> None:
         """
         Adds or updates a replacement for a symbol.
         Use this instead of `self.replacements[a] = expr`.
         """
+
+        # With unbacked symbols, it is not necessarily always profitable to do
+        # a replacement.  We may end up with a worse value range if we do the
+        # substitution, especially if the unbacked symbol to be replaced was
+        # marked as size like.
+        #
+        # For example let's suppose that 'a' in [2, inf].  If we replace 'a'
+        # with '10 - y', where y in [2, inf], we get a much worse bound.  This
+        # is bad, don't do the replacement!
+        new_bound = None
+        old_bound = None
+        if self.is_unbacked_symint(a):
+            env = {x: self.var_to_range.get(x, None) for x in expr.free_symbols}
+
+            # If you have x in [2, maxint], then 2*x in [4, 2*maxint].
+            # But we don't really care that the max bound says we can
+            # go beyond the maximum integer size, because we aren't
+            # using bigints anyway.  Arguably, ValueRanges should know
+            # to do this truncation automaticaly (to avoid doing
+            # bigint compute in range analysis), but right now it doesn't
+            # so we need to get rid of some unnecessary precision.
+            int_range = ValueRanges(-sys.maxsize - 1, sys.maxsize - 1)
+
+            new_bound = bound_sympy(expr, env) & int_range
+            old_bound = self.var_to_range[a] & int_range
+            if not new_bound.issubset(old_bound):
+                self.log.debug("skipped set_replacement %s = %s (%s) [%s not subset of %s]", a, expr, msg, new_bound, old_bound)
+                return
```

Comment:
This should be 
```suggestion
            self.var_to_range[a] = bound_sympy(expr, env) & self.var_to_range[a]
```
In particular, we should always do the replacement. There's no reason why we shouldn't.

We should update the comments and even the PR title.

Reply:
Well, we can't *always* do the replacement. But we always can update the VR (which I will do)

Reply:
why can't we always do the replacement?

Reply:
Sorry, not "can't", we don't always want to do the replacement if it makes the overall range worse.

Reply:
but if we are taking the intersection of both ranges it's never going to be worse, is it?

Reply:
When we do a replacement, we no longer can directly intersect the ranges. Specifically, to do the replacement y = f(x), the value range on y is no longer considered, and it may not be possible to update the value range on x so that it is tight with respect to the old range on y.